cgen/sycl.py

```diff
@@ -0,0 +1,240 @@
+                f"({self.item}.get_group())")
+
+    mapper_method = "map_sycl_local"
+
+
+class SYCLGlobal(DeclSpecifier):
+    def __init__(self, subdecl):
+        self.subdecl = subdecl
+
+    def get_decl_pair(self):
+        sub_tp, sub_decl = self.subdecl.get_decl_pair()
+        return [f"sycl::global_ptr<{sub_tp[0]}>"], sub_decl
+
+    mapper_method = "map_sycl_global"
+
+
+# TODO sycl Image
+class SYCLImage(Value):
+    def __init__(self, dims, mode, name):
+        if mode == "r":
+            spec = "__read_only"
+        elif mode == "w":
+            spec = "__write_only"
+        else:
+            raise ValueError("mode must be one of 'r' or 'w'")
+
+        Value.__init__(self, f"{spec} image{dims}d_t", name)
+
+    mapper_method = "map_cl_image"
+
+
+# }}}
+
+
+# {{{ function attributes
+class _SYCLWorkGroupSizeDeclarator(Generable):
+    def __init__(self, dim):
+
+        while len(dim) < 3:
+            dim = dim + (1,)
+        self.dim = dim
+        self.decl = 'extern "C" [[sycl::reqd_work_group_size({})]]'.format(
+            ", ".join(str(d) for d in self.dim)
+        )
+
+    def generate(self):
+        yield self.decl
+
+
+class SYCLRequiredWorkGroupSize(_SYCLWorkGroupSizeDeclarator):
+    mapper_method = "map_SYCL_required_wokgroup_size"
+
+
+# }}}
+
+# TODO SYCL vector
+# {{{ vector PODs
+
+# }}}
+class SYCLAccessor(NestedDeclarator):
+    def __init__(self, subdecl, handler, count=None):
+        NestedDeclarator.__init__(self, subdecl)
+        self.handler = handler
+        self.count = count
+        sub_tp, sub_decl = subdecl.get_decl_pair()
+        self.type = sub_tp[0]
+        self.sub_decl = sub_decl
+
+    def get_decl_pair(self):
+        if self.count is None:
+            return [f"sycl::accessor<{ self.type}>"
+                    ], f"{self.sub_decl}({self.handler})"
+        else:
+            return [
+                f"sycl::accessor<{ self.type}>"
+            ], f"{self.sub_decl}(sycl::range<1>({self.count}),{self.handler})"
+
+    def get_type(self):
+        return self.type
+
+    mapper_method = "map_sycl_accessor"
+
+
+class SYCLBody(Generable):
+    def __init__(self, body, ndim, handler, nd_item, queue, ndrange):
+        self.body = SYCLQueueSubmit(
+            Block([SYCLParallelFor(body, ndim, handler, nd_item, ndrange)]),
+            handler,
+            queue,
+        )
+
+    def generate(self):
+        yield from self.body.generate()
+
+    mapper_method = "map_SYCL_function_body"
+
+
+class SYCLParallelFor(Generable):
+    def __init__(self, body, ndim, handler, nd_item, ndrange):
+        self.handler = handler
+        self.args = [
+            ndrange,
+            FunctionBody(Lamda("=",
+                               [Value(f"sycl::nd_item<{ndim}>", nd_item)]), body),
+        ]
+
+    def generate(self):
+        yield "{}.parallel_for({});".format(
+            self.handler, ", ".join(str(ad) for ad in self.args)
+        )
```

Comment:
Is this special syntax? This looks like a normal expression, and expressions are out of scope for cgen. (See, e.g. pymbolic.)